compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java

```diff
@@ -2462,6 +2464,100 @@ public void visit(BLangSimpleVariable varNode) {
         varNode.annAttachments.forEach(annotationAttachment -> analyzeNode(annotationAttachment, env));
     }
 
+    private boolean isValidInferredArray(BLangNode node, List<BLangExpression> indexSizes) {
+        switch (node.getKind()) {
+            case INTERSECTION_TYPE_NODE:
+            case UNION_TYPE_NODE:
+                return isValidInferredArray(node.parent, indexSizes);
+            case VARIABLE:
+                BLangSimpleVariable varNode = (BLangSimpleVariable) node;
+                BLangExpression expr = varNode.expr;
+                return validateContext(node.parent) && expr != null && expr.getKind() == NodeKind.LIST_CONSTRUCTOR_EXPR
+                        && validateInitializer(indexSizes, indexSizes.size() - 2, (BLangListConstructorExpr) expr);
+            default:
+                return false;
+        }
+    }
+
+    private boolean validateContext(BLangNode node) {
+        switch (node.getKind()) {
+            case PACKAGE:
+            case EXPR_FUNCTION_BODY:
+            case BLOCK_FUNCTION_BODY:
+            case BLOCK:
+                return true;
+            case VARIABLE_DEF:
+                return validateContext(node.parent);
+            default:
+                return false;
+        }
+    }
+
+    private void getPositionsOfInferredArrays(BLangNode node, List<Location> posArray) {
+        switch (node.getKind()) {
+            case ARRAY_TYPE:
+                BLangArrayType arrayType = (BLangArrayType) node;
+                if (isInferredArray(arrayType.sizes)) {
+                    posArray.add(arrayType.pos);
+                }
+                return;
+            case INTERSECTION_TYPE_NODE:
+                BLangIntersectionTypeNode intersectionTypeNode = (BLangIntersectionTypeNode) node;
+                for (BLangType member : intersectionTypeNode.constituentTypeNodes) {
+                    getPositionsOfInferredArrays(member, posArray);
+                }
+                return;
+            case UNION_TYPE_NODE:
+                BLangUnionTypeNode unionTypeNode = (BLangUnionTypeNode) node;
+                for (BLangType member : unionTypeNode.memberTypeNodes) {
+                    getPositionsOfInferredArrays(member, posArray);
+                }
+                return;
+            default:
+                return;
+        }
+    }
+
+    private boolean validateInitializer(List<BLangExpression> indexSizes, int index, BLangListConstructorExpr expr) {
+        if (index < 0) {
+            return true;
+        }
+
+        boolean isInferred = false;
+        if ((indexSizes.get(index).getKind() == LITERAL) &&
```

Comment:
Shall we introduce variables for these get method returns and reuse them. eg: `indexSizes.get(index)`. Here this may be find because this is actually an ArrayList but in future if the data structure chagnes get method call may become expensive